devicemonitoringserver/commandcenter.cpp

```
@@ -0,0 +1,77 @@
+#include "commandcenter.h"
+
+#include "messagecommand.h"
+#include "messageerror.h"
+#include "messagemeterage.h"
+
+#include <algorithm>
+#include <cmath>
+#include <numeric>
+
+void CommandCenter::setSchedule(const DeviceWorkSchedule& schedule)
+{
+    m_scheduleInfo[schedule.deviceId] = {};
+    auto& scheduleInfo = m_scheduleInfo[schedule.deviceId];
+    scheduleInfo.phases = schedule.schedule;
+    std::sort(scheduleInfo.phases.begin(),
+              scheduleInfo.phases.end(),
+              [](const Phase& phaseA, const Phase& phaseB) { return phaseA.timeStamp < phaseB.timeStamp; });
+}
+
+void CommandCenter::processMeterage(uint64_t deviceId, MessageMeterage meterage, std::function<void(uint64_t, const Message&)> callback)
+{
+    auto currentTimeStamp = meterage.timeStamp();
+
+    auto lastTimeStampIt = m_lastTimeStamp.find(deviceId);
+    if (lastTimeStampIt != m_lastTimeStamp.end() && lastTimeStampIt->second >= currentTimeStamp)
+    {
+        callback(deviceId, MessageError(MessageError::ErrorType::Obsolete));
+        return;
+    }
+    m_lastTimeStamp[deviceId] = currentTimeStamp;
```

Может лучше только для устройств, для которых задан план, запоминать какие-то параметры?

---

Я тут подумал про принцип Fail Fast: если устройство шлет сообщения с неверными временными метками, лучше ему сразу об этом сообщить, чем дожидаться выставления плана для этого устройства. Но если цель - экономия ресурсов сервера, можно, конечно, этого не делать.

---

Понял тебя